File: tstock.c
// Gabe Banks
// 1/3/22

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <ctype.h>
#include <wchar.h>
#include <locale.h>
#include <time.h>
#include <math.h>
#include <unistd.h>

#include <curl/curl.h>
#include <curl/urlapi.h>

// Size of the buffer used for curl requests
#define CURL_RAW_BUFF_SIZE 20000

// Function declarations
void printHelp();
double map(double x, double l1, double h1, double l2, double h2);
// Gets the raw data from the MarketStack API.
char *getMarketStackData(const char *apiKey, const struct tm *dateFrom, const struct tm *dateTo, const char *ticker);
// Simple CURL callback to append data to a buffer
static size_t appendToBufferWriter(char *data, size_t size, size_t nmemb, char *buffer);
int main(int argc, char *argv[]);

// Print the help message and exit
void printHelp()
{
    printf("Usage: tstock [OPTIONS]... TICKER\n"
           "Prints a candlestick chart of TICKER in the terminal.\n"
           "Options:\n"
           "    -d [days]       Number of days to go back in API call. Defaults to 90.\n"
           "    -h              print this message and exit\n"
           "    -v              enables verbosity\n"
           "    -y [lines]      Specify height of the chart. Defaults to 40.\n");
    return;
}

// Maps a number x from scale l1, h1 to relative scale in l2, h2
double map(double x, double l1, double h1, double l2, double h2)
{
    return (((x - l1) / (h1 - l1)) * (h2 - l2)) + l2;
}

static size_t appendToBufferWriter(char *data, size_t size, size_t nmemb, char *buffer)
{
    if (!buffer)
        return (0);
    // TODO: Does this work if there is a \0 in the data? Can there be a \0 in the data?
    if (strlen(buffer) + size * nmemb > CURL_RAW_BUFF_SIZE)
        exit(EXIT_FAILURE);
    strncat(buffer, data, size * nmemb);
    return size * nmemb;
}

char *getMarketStackData(const char *apiKey, const struct tm *dateFrom, const struct tm *dateTo, const char *ticker)
{
    // We allocate memory for the data returned by the request
    char *curlBuffer = malloc(CURL_RAW_BUFF_SIZE);
    memset(curlBuffer, 0, CURL_RAW_BUFF_SIZE);
    CURL *curl = curl_easy_init();

    if (!curl)
        return (0);

    // TODO: Handle errors from curl_easy_setopt

    // Set curl buffer
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, appendToBufferWriter);
    curl_easy_setopt(curl, CURLOPT_WRITEDATA, curlBuffer);

    // Generate the API url
    CURLU *url = curl_url();
    curl_url_set(url, CURLUPART_URL, "http://api.marketstack.com/v1/eod", 0);

    char tokenParam[11 + 32 + 1] = "access_key="; // strlen("access_key=") == 11 && strlen("<MARKETSTACK_API_KEY>") == 32
    strcat(tokenParam, apiKey);
    curl_url_set(url, CURLUPART_QUERY, tokenParam, CURLU_APPENDQUERY);

    // TODO: Maybe write a small string view struct for easier string manipulation? This is... not great...
    char dateFromParam[10 + 10 + 1] = "date_from="; // The dates will always be 10 characters long with the provided format, and "date_from=" is also 10 characters
    char dateFromRaw[10 + 1];
    strftime(dateFromRaw, sizeof(dateFromRaw), "%Y-%m-%d", dateFrom);
    strcat(dateFromParam, dateFromRaw);
    curl_url_set(url, CURLUPART_QUERY, dateFromParam, CURLU_APPENDQUERY);

    char dateToParam[10 + 8 + 1] = "date_to="; // The dates will always be 10 characters long with the provided format, and "date_to=" is 8 characters
    char dateToRaw[10 + 1];
    strftime(dateToRaw, sizeof(dateToRaw), "%Y-%m-%d", dateTo);
    strcat(dateToParam, dateToRaw);
    curl_url_set(url, CURLUPART_QUERY, dateToParam, CURLU_APPENDQUERY);

    char symbolsParam[32] = "symbols=";
    strcat(symbolsParam, ticker);
    curl_url_set(url, CURLUPART_QUERY, symbolsParam, CURLU_APPENDQUERY);

    char *finalUrl;
    curl_url_get(url, CURLUPART_URL, &finalUrl, 0);
    curl_url_cleanup(url);

    // TODO: Check if verbose flag is set to log
    fprintf(stdout, "Getting data from: %s\n", finalUrl);

    curl_easy_setopt(curl, CURLOPT_URL, finalUrl);

    CURLcode res = curl_easy_perform(curl);
    (void) res; // TODO: Check res for errors
    curl_easy_cleanup(curl);
    return (curlBuffer);
}

int main(int argc, char *argv[])
{
    // Set locale for unicode characters
    setlocale(LC_CTYPE, "");

    // maxX, maxY determine the dimensions, in characters, of the graph
    int maxX;      // This is set to the number of candles + 12
    int maxY = 40; // Can be set with -y

    // Variables that can be modified with flags
    int verbose = 0;   // Toggled with -v
    int daysBack = 90; // Set with -d
    char strDaysBack[10] = "90";

    // Index variables for loops
    int i = 0;
    int j = 0;

    // Parse options
    int opt; // TODO: setup args for different API backends
    while ((opt = getopt(argc, argv, ":d:y:vh")) != -1)
    {
        switch (opt)
        {
        case 'd':
            strcpy(strDaysBack, optarg);
            daysBack = atoi(optarg);
            break;
        case 'y':
            maxY = atoi(optarg);
            break;
        case 'h':
            printHelp();
            return 0;
            break;
        case 'v':
            verbose = 1;
            break;
        case ':':
            printf("Option -%c needs a value.\n", optopt);
            printHelp();
            return 0;
            break;
        case '?':
            printf("Unknown option `-%c'.\n", optopt);
            printHelp();
            return 0;
            break;
        }
    }
    // Check for right number of arguments
    if (optind != argc - 1)
    {
        printHelp();
        return 1;
    }

    // Stores the ticker provided as argument by user
    char ticker[50];
    strcpy(ticker, argv[optind]);

    // Check for API key stored in env variable $MARKETSTACK_API_KEY
    const char *apikey = getenv("MARKETSTACK_API_KEY");
    if (apikey == NULL)
    {
        printf("API key not detected! Follow these instructions to get your API Key working:\n");
        printf("- Make a free MarketStack API account at https://marketstack.com/signup/free\n");
        printf("- Login and find your API Access Key on the Dashboard page\n");
        printf("- Run \"export MARKETSTACK_API_KEY=<your access key>\".\n");
        printf("    You can make this permanent by adding a line like \"export MARKETSTACK_API_KEY=<your access key>\" to your .bashrc\n");
        return (1);
    }

    // Find date values
    time_t now = time(NULL);
    // Here we dereference the value from localtime() to copy it so that the second call to localtime() doesn't overwrite the data from the first
    struct tm tm_now = *localtime(&now);
    time_t start = time(NULL) - (86400 * daysBack); // defaults to 90 days * seconds in a day (86400)
    struct tm tm_start = *localtime(&start);

	// This will malloc some memory, we need to free `data` once we're done!
    char *data = getMarketStackData(apikey, &tm_start, &tm_now, ticker);

    // If the received data is under 100 characters, something probably went wrong
    if (strlen(data) < 100)
    {
        printf("Didn't receive enough data from API. Check your internet connection. Is this stock covered by the API?\n");
        if (verbose)
            printf("Received following data from API: %s\n", data);
        return (1);
    }

    if (verbose)
        printf("Raw JSON data received:\n%s\nParsing...\n", data);

    // Code for parsing JSON data:
    // looking for either 0=open, 1=high, 2=low, 3=close, 4=day
    int looking = 0;
    // Temporary location to store variables when parsing raw API data
    char tempPriceStr[500];
    double tempPrice;
    // Temporary location for date while parsing
    char tmpDay[4];
    // Final location for parsed bar data.
    // Right now just pray that the user doesn't try to pull over 1000 days worth of data...
    // Each row is of the shape: [Open, high, low, close, day]
    double barData[1000][5];
    // Index variable
    int bari = 0;
    // ath, atl, for keeping track of all-time-high and all-time-low prices, within the specified time window.
    double ath = 0;
    double atl = 99999999;

    // Start parsing API data
    i = 0;
    while (data[i] != ']')
    {
        if (data[i - 1] == '"' && data[i] == 'e' && data[i + 1] == 'r' && data[i + 2] == 'r' && data[i + 3] == 'o' && data[i + 4] == 'r')
        {
            printf("ERROR: The API returned an error: %s\n", data);
            return (1);
        }
        switch (looking)
        {
        case 0:
            if (data[i - 1] == '"' && data[i] == 'o' && data[i + 1] == 'p' && data[i + 2] == 'e' && data[i + 3] == 'n')
            {
                looking = 1;
                j = i + 6;
                while (data[j] != ',')
                {
                    tempPriceStr[j - i - 6] = data[j];
                    j++;
                }
                tempPriceStr[j - i - 5] = '\0';
                tempPrice = atof(tempPriceStr);
                barData[bari][0] = tempPrice;
            }
            break;
        case 1:
            if (data[i - 1] == '"' && data[i] == 'h' && data[i + 1] == 'i' && data[i + 2] == 'g' && data[i + 3] == 'h')
            {
                looking = 2;
                j = i + 6;
                while (data[j] != ',')
                {
                    tempPriceStr[j - i - 6] = data[j];
                    j++;
                }
                tempPriceStr[j - i - 5] = '\0';
                tempPrice = atof(tempPriceStr);
                barData[bari][1] = tempPrice;
                if (tempPrice > ath)
                {
                    ath = tempPrice;
                }
            }
            break;
        case 2:
            if (data[i - 1] == '"' && data[i] == 'l' && data[i + 1] == 'o' && data[i + 2] == 'w')
            {
                looking = 3;
                j = i + 5;
                while (data[j] != ',')
                {
                    tempPriceStr[j - i - 5] = data[j];
                    j++;
                }
                tempPriceStr[j - i - 4] = '\0';
                tempPrice = atof(tempPriceStr);
                barData[bari][2] = tempPrice;
                if (tempPrice < atl)
                {
                    atl = tempPrice;
                }
            }
            break;
        case 3:
            if (data[i - 1] == '"' && data[i] == 'c' && data[i + 1] == 'l' && data[i + 2] == 'o' && data[i + 3] == 's' && data[i + 4] == 'e')
            {
                looking = 4;
                j = i + 7;
                while (data[j] != ',')
                {
                    tempPriceStr[j - i - 7] = data[j];
                    j++;
                }
                tempPriceStr[j - i - 6] = '\0';
                tempPrice = atof(tempPriceStr);
                barData[bari][3] = tempPrice;
            }
            break;
        case 4:
            if (data[i - 1] == '"' && data[i] == 'd' && data[i + 1] == 'a' && data[i + 2] == 't' && data[i + 3] == 'e')
            {
                looking = 0;
                tmpDay[0] = data[i + 15];
                tmpDay[1] = data[i + 16];
                tmpDay[2] = '\0';
                barData[bari][4] = atof(tmpDay);
                bari++;
            }
            break;

        default:
            printf("ERROR: looking has an invalid value\n");
            return (1);
        }
        i++;
    }

    // We're no longer using the data, so we can free the memory back
    free(data);

    // The final value of bari will be the number of rows barData has. Store in barYLen
    int barYLen = bari;
    if (verbose)
        printf("Found %i days of OHLC data.\n", barYLen);
    // Reverse the barData, so that earliest is first
    double barDataRev[1000][5];
    for (i = 0; i < barYLen; i++)
    {
        for (j = 0; j < 5; j++)
        {
            barDataRev[i][j] = barData[barYLen - i - 1][j];
        }
    }
    // Set X dimension of graph to number of candles + 12
    maxX = barYLen + 12;

    // Create 2d array of unicode characters that will become the graph
    wchar_t graph[maxY][maxX];

    // Clear the array by setting every element to a blank space
    for (i = 0; i < maxY; i++)
    {
        for (j = 0; j < maxX; j++)
        {
            graph[i][j] = ' ';
        }
    }

    // Draw top and bottom borders
    for (i = 0; i < maxX; i++)
    {
        graph[0][i] = 0x2500;                              // 0x2500 = ─
        graph[maxY - 1][i] = i % 5 == 0 ? 0x253c : 0x2500; // every fifth character, draw a "┼"
    }
    // Draw left and right borders
    for (i = 0; i < maxY; i++)
    {
        graph[i][0] = i % 5 == 0 ? 0x253c : 0x2502; // every fifth character, draw a "┼"
        graph[i][maxX - 1] = 0x2502;
    }
    // Draw corners
    graph[0][0] = 0x250c;               // 0x250c = ┌
    graph[0][maxX - 1] = 0x2510;        // 0x2510 = ┐
    graph[maxY - 1][0] = 0x2514;        // 0x2514 = └
    graph[maxY - 1][maxX - 1] = 0x2518; // 0x2518 = ┘

    // Don't even bother with graph title if there's not 40 days worth of data
    if (daysBack >= 40)
    {
        // Draw graph title
        char upper[10];
        for (i = 0; i < strlen(ticker); i++)
        {
            upper[i] = toupper(ticker[i]);
        }
        upper[i] = '\0';
        char title[50] = "  ";
        strcat(title, strDaysBack);
        strcat(title, " Day Stock Price for $");
        strcat(title, upper);
        strcat(title, "  ");
        graph[0][1] = 0x2524; // ┤
        for (i = 0; i < strlen(title); i++)
        {
            graph[0][i + 2] = title[i];
        }
        graph[0][i + 2] = 0x251c; // ├
    }

    // Declare variables for drawing candlesticks
    // Lower and upper bounds of which rows to draw a candle stick
    int low = 4;
    int high = maxY - 4;
    // mlow, mhigh = mapped low and mapped high. Prices that have been mapped to the scale of the Y axis.
    int mlow;
    int mhigh;
    int vpos;
    // Temp variable for switching mlow and mhigh
    int tmp;
    // Set index to start at 5 for padding inside chart
    j = 5;
    // Declare array for keeping track of which columns saw an increase or a decrease in stock price
    int columnColors[maxX]; // 0 = white, 1 = red, 2 = green
    for (i = 0; i < maxX; i++)
    {
        columnColors[i] = 0; // Initialize by setting every element to 0
    }

    for (i = 0; i < barYLen; i++)
    {
        // map high/low parts of the candlestick
        mhigh = map(barDataRev[i][2], atl, ath, high, low);
        mlow = map(barDataRev[i][1], atl, ath, high, low);
        // fill in the high/low part of the candlestick
        for (vpos = mlow; vpos <= mhigh; vpos++)
        {
            graph[vpos][j] = 0x2502;
        }
        // map open/close parts of the candlestick
        mhigh = map(barDataRev[i][0], atl, ath, high, low);
        mlow = map(barDataRev[i][3], atl, ath, high, low);
        // If "mlow" is greater than "mhigh", that means the stock went down in price
        if (mlow > mhigh)
        {
            tmp = mlow;
            mlow = mhigh;
            mhigh = tmp;
            columnColors[j] = 1;
        }
        else
        { // Otherwise the stock increased in price
            columnColors[j] = 2;
        }
        // fill in the open/close part of the candlestick
        for (vpos = mlow; vpos <= mhigh; vpos++)
        {
            graph[vpos][j] = 0x2588;
        }
        j++;
    }

    // Print graph to terminal
    // String holding an axis label
    char label[15];
    sprintf(label, "$%.2f", ath);
    // Margin between the y-axis of the graph and the left edge of the screen
    int margin = strlen(label);
    // Variable for holding a numeric stock price
    double price;
    // Start looping through "graph" and printing character by character
    for (i = 0; i < maxY; i++)
    {
        // ANSI escape code for resetting font/color
        printf("\x1b[0m");
        // Print y axis labels and margin
        if (i % 5 == 0 && i >= low && i <= high)
        {
            price = map(i, low, high, ath, atl);
            sprintf(label, "$%.2f", price);
            for (j = 0; j < margin - strlen(label); j++)
            {
                printf(" ");
            }
            printf("$%.2f", price);
        }
        else
        {
            for (j = 0; j < margin; j++)
            {
                printf(" ");
            }
        }
        // Print graph contents
        for (j = 0; j < maxX; j++)
        {
            if (i != 0 && i != maxY - 1)
            {
                switch (columnColors[j])
                { // ANSI Escape sequences for setting print color
                case 0:
                    printf("\x1b[0m"); // White
                    break;
                case 1:
                    printf("\x1b[31m"); // Red
                    break;
                case 2:
                    printf("\x1b[32m"); // Green
                    break;

                default:
                    printf("ERROR: Incorrect columncolor\n");
                    return (1);
                }
            }
            printf("%lc", graph[i][j]);
        }
        printf("\n");
    }
    for (j = 0; j < margin; j++)
    {
        printf(" ");
    }
    printf("     "); // account for margin
    // Print x axis labels
    for (j = 0; j < barYLen; j++)
    {
        if (j % 5 == 0)
        {
            printf("%i", (int)barDataRev[j][4]);
            if ((int)barDataRev[j][4] > 9)
            {
                printf("   ");
            }
            else
            {
                printf("    ");
            }
        }
    }

    printf("\n\n");
    return (0);
}

File: tstock.py
#!/usr/bin/python

import os
import argparse
import datetime
from xmlrpc.client import MAXINT
import requests
import numpy as np

def translate(x, l1, h1, l2, h2):
    return (((x - l1) / (h1 - l1)) * (h2 - l2)) + l2

def get_api_key():
    if not 'MARKETSTACK_API_KEY' in list(os.environ.keys()):
        print("error: API key not detected! Follow these instructions to get your API Key working:\n" + \
        "- Make a free MarketStack API account at https://marketstack.com/signup/free\n" + \
        "- Login and find your API Access Key on the Dashboard page\n" + \
        "- Run \"export MARKETSTACK_API_KEY=<your access key>\"." + \
        "You can make this permanent by adding this line to your .bashrc\n")
        exit(1)
    return os.environ['MARKETSTACK_API_KEY']


def main():
    padX = 5 # TODO: make padX an option
    padY = 4
    days_back = 90
    ticker = "aapl"
    verbose = True
    maxY = 40

    # HTTP GET API data
    apikey = get_api_key()
    date_to = datetime.datetime.today().strftime('%Y-%m-%d')
    date_from = (datetime.datetime.today() -
                datetime.timedelta(days=days_back)).strftime('%Y-%m-%d')
    request_url = f'http://api.marketstack.com/v1/eod?access_key={apikey}&date_from={date_from}&date_to={date_to}&symbols={ticker}'

    if verbose:
        print(f"Days Back: {days_back}\nTicker: {ticker}\nAPI Key: {apikey}\nRequest URL: {request_url}")

    r = requests.get(request_url).json()
    if 'error' in list(r.keys()):
        print(f"error: The API returned the following error:\n{r}")
        exit(1)
    data = r['data']

    # Parse API data
    candlesticks = []
    for d in data:
        candlesticks.append((d['open'], d['high'], d['low'], d['close'], d['date'][8:10]))

    candlesticks = list(reversed(candlesticks))
    # candlesticks = list(reversed([(171.34, 173.78, 171.09, 173.07, '14'), (175.78, 176.62, 171.79, 172.19, '13'), (176.12, 177.179, 174.821, 175.53, '12'), (172.32, 175.18, 170.82, 175.08, '11'), (169.08, 172.5, 168.17, 172.19, '10'), (172.89, 174.14, 171.03, 172.17, '07'), (172.7, 175.3, 171.64, 172.0, '06'), (179.61, 180.17, 174.64, 174.92, '05'), (182.63, 182.9, 179.12, 179.7, '04'), (177.83, 182.88, 177.71, 182.01, '03'), (178.085, 179.23, 177.26, 177.57, '31'), (179.47, 180.57, 178.09, 178.2, '30'), (179.33, 180.63, 178.14, 179.38, '29'), (180.16, 181.33, 178.53, 179.29, '28'), (177.085, 180.42, 177.07, 180.33, '27'), (175.85, 176.85, 175.27, 176.28, '23'), (173.04, 175.86, 172.15, 175.64, '22'), (171.56, 173.2, 169.12, 172.99, '21'), (168.28, 170.58, 167.46, 169.75, '20'), (169.93, 173.47, 169.69, 171.14, '17'), (179.28, 181.14, 170.75, 172.26, '16'), (175.11, 179.5, 172.31, 179.3, '15'), (175.25, 177.74, 172.21, 174.33, '14'), (181.115, 182.09, 175.53, 175.74, '13'), (175.205, 179.63, 174.69, 179.45, '10'), (174.91, 176.75, 173.92, 174.56, '09'), (172.125, 175.95, 170.7, 175.08, '08'), (169.08, 171.58, 168.34, 171.18, '07'), (164.29, 167.8799, 164.28, 165.32, '06'), (164.02, 164.96, 159.72, 161.84, '03'), (158.74, 164.2, 157.8, 163.76, '02'), (167.48, 170.295, 164.53, 164.77, '01'), (159.99, 165.52, 159.92, 165.3, '30'), (159.37, 161.19, 158.7901, 160.24, '29'), (159.57, 160.45, 156.36, 156.81, '26'), (160.75, 162.14, 159.64, 161.94, '24'), (161.12, 161.8, 159.07, 161.41, '23'), (161.68, 165.7, 161.0, 161.02, '22'), (157.65, 161.02, 156.53, 160.55, '19'), (153.71, 158.67, 153.05, 157.87, '18'), (150.995, 155.0, 150.995, 153.49, '17'), (149.94, 151.49, 149.34, 151.0, '16'), (150.37, 151.88, 149.43, 150.0, '15'), (148.43, 150.4, 147.48, 149.99, '12'), (148.96, 149.43, 147.68, 147.87, '11'), (150.02, 150.1297, 147.85, 147.92, '10'), (150.2, 151.428, 150.07, 150.81, '09'), (151.41, 151.55, 150.16, 150.44, '08'), (151.89, 152.2, 150.06, 151.28, '05'), (151.58, 152.43, 150.64, 150.96, '04'), (150.39, 151.97, 149.83, 151.49, '03'), (148.66, 151.57, 148.65, 150.02, '02'), (148.985, 149.7, 147.8, 148.96, '01'), (147.22, 149.94, 146.41, 149.8, '29'), (149.82, 153.165, 149.72, 152.57, '28'), (149.36, 149.73, 148.49, 148.85, '27'), (149.33, 150.84, 149.01, 149.32, '26'), (148.68, 149.37, 147.6211, 148.64, '25'), (149.69, 150.18, 148.64, 148.69, '22'), (148.81, 149.64, 147.87, 149.48, '21'), (148.7, 149.7522, 148.12, 149.26, '20'), (147.01, 149.17, 146.55, 148.76, '19'), (143.445, 146.84, 143.16, 146.55, '18')]))
    maxX = len(candlesticks) + padX * 2 + 2


    # Create the chart
    chart = np.array([[" " for x in range(maxX)] for y in range(maxY)])
    column_colors = ["\x1b[0m" for x in range(maxX)] # Stores ANSI escape sequences for printing color
    # Draw borders
    chart[0, :] = "─"
    chart[-1, :] = "─"
    chart[:, 0] = "│"
    chart[:, -1] = "│"
    chart[0, 0] = "┌"
    chart[0, -1] = "┐"
    chart[-1, 0] = "└"
    chart[-1, -1] = "┘"
    # Draw graph title, if there are there enough worth of data to contain it
    title = f"┤  {days_back} Day Stock Price for ${ticker.upper()}  ├"
    if maxX >= len(title) + 2:
        for i, c in enumerate(title):
            chart[0, i+1] = c
    # Find all time high and all time low
    ath = 0
    atl = MAXINT
    for c in candlesticks:
        if c[1] > ath:
            ath = c[1]
        if c[2] < atl:
            atl = c[2]
    print(ath, atl)
    # Draw candlesticks
    start_i = 1 + padX
    end_i = maxX - 1 - padX
    y_axis_low = padY
    y_axis_high = maxY - padY
    print(y_axis_high)
    for i, c in enumerate(candlesticks):
        shifted_i = i+start_i
        # Stuff gets a little confusing here because the graph has to be y-inverted. "high" is referring to a high price, but needs to be flipped to a low index.
        translated_open = int(translate(c[0], atl, ath, y_axis_high, y_axis_low))
        translated_high = int(translate(c[1], atl, ath, y_axis_high, y_axis_low))
        translated_low = int(translate(c[2], atl, ath, y_axis_high, y_axis_low))
        translated_close = int(translate(c[3], atl, ath, y_axis_high, y_axis_low))
        # Draw high/low
        for y in range(translated_high, translated_low+1):
            chart[y, shifted_i] = "|"
        # Draw open/close
        # Positive day, stock went up
        if c[0] < c[3]:
            column_colors[shifted_i] = "\x1b[32m" # ANSI green
            tmp = translated_low
            translated_low = translated_high
            translated_high = tmp
            for y in range(translated_close, translated_open+1):
                chart[y, shifted_i] = "█"
        # Negative day, stock went down
        else:
            column_colors[shifted_i] = "\x1b[31m" # ANSI red
            for y in range(translated_open, translated_close+1):
                chart[y, shifted_i] = "█"

    # Setup x-axis labels
    x_axis_labels = " " * (1+padX)
    for i in range(start_i, end_i):
        shifted_i = i-start_i
        if (shifted_i) % 5 == 0:
            chart[-1, i] = "┼"
            if int(candlesticks[shifted_i][4]) >= 10:
                x_axis_labels += f"{candlesticks[shifted_i][4]}   "
            else:
                x_axis_labels += f"{int(candlesticks[shifted_i][4])}    "
    x_axis_labels += " " * (maxX - len(x_axis_labels))

    # Setup y-axis labels
    y_axis_labels = []
    margin = len("${:,.2f}".format(ath))
    for i in range(maxY):
        if i >= y_axis_low and i <= y_axis_high:
            shifted_i = y_axis_high - i
            if shifted_i % 4 == 0:
                chart[i, 0] = "┼"
                label = "${:,.2f}".format(translate(shifted_i, y_axis_low, y_axis_high, atl, ath))
                y_axis_labels.append(" " * (margin - len(label)) + f"{label}")
            else:
                y_axis_labels.append(" " * margin)
        else:
            y_axis_labels.append(" " * margin)

    # Print out the chart
    for y, row in enumerate(chart):
        out = ""
        out += y_axis_labels[y]
        for x, char in enumerate(row):
            if y >= y_axis_low and y <= y_axis_high:
                out += column_colors[x]
            out += char
        print(out)
    # Print x axis labels
    print(y_axis_labels[0] + x_axis_labels)

if __name__ == "__main__":
    main()